File: python/aln/router.py
from datetime import datetime
from threading import Lock, Thread
import random
import time
from .packet import Packet, readINT16U, writeINT16U

bytesToAddressType = str

def makeNetQueryPacket():
    return Packet(netState=Packet.NET_QUERY)

def byteLen(buff):
    return len(buff).to_bytes(1, 'big')

def makeNetworkRouteSharePacket(srcAddr, destAddr, cost):
    data = []
    data.extend(byteLen(destAddr))
    data.extend(bytes(destAddr, "utf-8"))
    data.extend(writeINT16U(cost))
    return Packet(netState=Packet.NET_ROUTE, srcAddr=srcAddr,data=data)

def parseNetworkRouteSharePacket(packet): # returns dest, next-hop, cost, err
    if packet.netState != Packet.NET_ROUTE:
        return (0, 0, 0, "parseNetworkRouteSharePacket: packet.NetState != NET_ROUTE")
    addrSize = int(packet.data[0])
    if len(packet.data) != (addrSize + 3):
        return (0, 0, 0, "parseNetworkRouteSharePacket: len(packet.Data) != (addrSize + 3)")
    addr = bytearray(packet.data[1:addrSize+1]).decode("utf-8")
    cost = readINT16U(packet.data[addrSize+1:addrSize+3])
    return addr, packet.srcAddr, cost, None

def makeNetworkServiceSharePacket(hostAddr, service, serviceCapacity):
    data = []
    data.extend(byteLen(hostAddr))
    data.extend(bytes(hostAddr, "utf-8"))
    data.extend(byteLen(service))
    data.extend(bytes(service, "utf-8"))
    data.extend(writeINT16U(serviceCapacity))
    return Packet(netState=Packet.NET_SERVICE, data=data)

# returns hostAddr, service, capacity, error
def parseNetworkServiceSharePacket(packet):
    if packet.netState != Packet.NET_SERVICE:
        return (0, 0, 0, "parseNetworkRouteSharePacket: packet.NetState != NET_ROUTE")

    if len(packet.data) < 4:
        return (0, 0, 0, "parseNetworkRouteSharePacket: payload too small")

    # TODO validate message length to avoid seg fault
    addrSize = int(packet.data[0])
    hostAddr = bytearray(packet.data[1:addrSize+1]).decode("utf-8")
    srvSize = int(packet.data[addrSize+1])
    service = bytearray(packet.data[addrSize+2:addrSize+2+srvSize]).decode("utf-8")
    capacity = readINT16U(packet.data[addrSize+2+srvSize:])
    return hostAddr, service, capacity, None

class RemoteNode:
    def __init__(self, address, nextHop, cost, channel, lastSeen=datetime.now()):
        self.address = address
        self.nextHop = nextHop
        self.cost = cost
        self.channel = channel
        self.lastSeen = lastSeen

class Router(Thread):
    def __init__(self, selector, address=0):
        super(Router, self).__init__()
        self.lock = Lock()
        self.selector = selector # top-level application event loop
        self.address = address   # network node ID for this router
        self.channels = []       # pool of all channels for flood propagation
        self.contextMap = {}     # serviceHandlerMap
        self.remoteNodeMap = {}  # RemoteNodeMap // map[address]RemoteNodes
        self.serviceMap = {}     # map[address]callback registered local service handlers
        self.serviceCapacityMap = {} # ServiceCapacityMap // map[service][address]capacity (remote services)
        self.serviceQueue = {}   # map[service]PacketList
        self.onServiceCapacityChanged = None # callback method (optional)
        self.stop = False
        self.daemon = True

    def run(self):
        while not self.stop:
            events = self.selector.select()
            for key, mask in events:
                callback = key.data
                callback(key.fileobj, mask)

    def handle_netstate(self, channel, packet):
        if packet.netState == Packet.NET_ROUTE: #  neighbor is sharing it's routing table')
            remoteAddress, nextHop, cost, err = parseNetworkRouteSharePacket(packet)
            if err is not None:
                print("err:",err)
            elif cost == 0: # remove route
                if remoteAddress == self.address: # readvertize self
                    time.sleep(0.1)
                    self.share_net_state()
                else: # remove route
                    # print("removing route to", remoteAddress)
                    del self.remoteNodeMap[remoteAddress]
                    for capacityMap in self.serviceCapacityMap.values():
                        del capacityMap[address]
                    for ch in self.channels:
                        if ch != channel:
                             ch.send(packet)
            else:
                # msg = "recvd NET_ROUTE to:[{rem}] via:[{next}] cost:{cost}"
                # print(msg.format(rem=remoteAddress, next=nextHop, cost=cost))
                if remoteAddress not in self.remoteNodeMap:
                    remoteNode = RemoteNode(remoteAddress, nextHop, cost, channel)
                    self.remoteNodeMap[remoteAddress] = remoteNode
                else:
                    remoteNode = self.remoteNodeMap[remoteAddress]
                    if remoteNode.channel not in self.channels or cost < remoteNode.cost or remoteNode.cost == 0:
                        remoteNode.nextHop = nextHop
                        remoteNode.channel = channel
                        remoteNode.cost = cost
                        #  relay update to other channels
                        p = makeNetworkRouteSharePacket(self.address, remoteAddress, cost+1)
                        for chan in self.channels:
                            if chan is not channel:
                                chan.send(p)
        elif packet.netState == Packet.NET_SERVICE: # neighbor is sharing service capacity info
            address, service, capacity, err = parseNetworkServiceSharePacket(packet)
            if err is not None:
                print("error parsing NET_SERVICE: {0}", err)
            elif address != self.address:
                if service not in self.serviceCapacityMap:
                    self.serviceCapacityMap[service] = {}
                
                if capacity == 0 and address in self.serviceCapacityMap[service]: # remove zero capacity services from the service map
                    del self.serviceCapacityMap[service][address]
                    self._on_service_capacity_changed(service, 0)

                if capacity != 0 and address not in self.serviceCapacityMap[service] or \
                    self.serviceCapacityMap[service][address] != capacity: # skip if no change
                    self.serviceCapacityMap[service][address] = capacity
                    # share the update with neighbors
                    for chan in self.channels:
                        if chan is not channel:
                            chan.send(packet)
                    self._on_service_capacity_changed(service, capacity)
                
        elif packet.netState == Packet.NET_QUERY:
            self.share_net_state()
    
    def on_packet(self, channel, packet):
        if packet.netState:
            with self.lock:
                self.handle_netstate(channel, packet)
        else:
            self.send(packet)

    def remove_channel(self, channel):
        with self.lock:
            self.channels.remove(channel)
            delete = [addr for addr in self.remoteNodeMap if self.remoteNodeMap[addr].channel == channel]
            for addr in delete:
                del self.remoteNodeMap[addr]
                for ch in self.channels:
                    ch.send(makeNetworkRouteSharePacket(self.address, addr, 0))

    def add_channel(self, channel):
        channel.on_close(self.remove_channel)
        with self.lock:
            self.channels.append(channel)
            channel.listen(self.selector, self.on_packet)
            channel.send(makeNetQueryPacket())
    
    def send(self, packet):
        if packet.srcAddr == None:
            packet.srcAddr = self.address

        # print("router.send from {0} to {1} via {2}, service:{3}, ctxID:{4}, datalen:{5}, data:{6}".format(
        #     packet.srcAddr, packet.destAddr, packet.nextAddr, packet.service, packet.contextID, len(packet.data), packet.data.decode("utf-8"))
        # )

        packetHandler = None
        if packet.destAddr is None and packet.service is not None:
            addresses = self.service_addresses(packet.service)
            if len(addresses):
                for i in range(1, len(addresses)):
                    pc = packet.copy()
                    pc.destAddr = addresses[i]
                    self.send(pc)
                packet.destAddr = addresses[0]
            else :
                return ("service of type " + packet.service + " not yet discovered on network")

        with self.lock:
            if packet.destAddr == self.address:
                if packet.service in self.serviceMap:
                    packetHandler = self.serviceMap[packet.service]
                elif packet.contextID in self.contextMap:
                    packetHandler = self.contextMap[packet.contextID]
                else:
                    return ("no suitable handler found for inbound packet")

            elif packet.nextAddr == self.address or packet.nextAddr == None:
                if packet.destAddr in self.remoteNodeMap:
                    route = self.remoteNodeMap[packet.destAddr]
                    packet.srcAddr = self.address
                    packet.nextAddr = route.nextHop
                    route.channel.send(packet)
                else:
                    return "no route for " + str(packet.destAddr)
            else:
                return "packet is unroutable; no action taken"
        if packetHandler:
            packetHandler(packet)
        return None

    def set_on_service_capacity_changed_handler(self, callback):
        self.onServiceCapacityChanged = callback
    
    def _on_service_capacity_changed(self, service, capacity):
        if self.onServiceCapacityChanged:
            self.onServiceCapacityChanged(service, capacity);

    def register_context_handler(self, callback):
        with self.lock:
            ctxID = random.randint(2, 65535)
            while ctxID in self.contextMap:
                ctxID = random.randint(2, 65535)
            self.contextMap[ctxID] = callback
            return ctxID

    def release_context(self, ctxID):
        with self.lock:
            self.contextMap.pop(ctxID, None)

    def register_service(self, service, handler):
        with self.lock:
            self.serviceMap[service] = handler
        self.export_services()

    def unregister_service(self, service):
        with self.lock:
            self.serviceMap.pol(service, None)
        # TODO send service offline packet
        self.share_net_state()

    def service_addresses(self, service):
        addresses = []
        with self.lock:            
            if service in self.serviceMap:
                addresses.append(self.address)
            if service in self.serviceCapacityMap: 
                for addr in self.serviceCapacityMap[service]:
                    addresses.append(addr)
        return addresses

    def export_routes(self):
        # compose routing table as an array of packets
        # one local route, with a cost of 1
        # for each remote route, our cost and add 1
        routes = [makeNetworkRouteSharePacket(self.address, self.address, 1)]
        for remoteAddress in self.remoteNodeMap:
            # TODO filter expired nodes by lastSeen 
            remoteNode = self.remoteNodeMap[remoteAddress]
            routes.append(makeNetworkRouteSharePacket(self.address, remoteAddress, remoteNode.cost+1))
        return routes

    def export_services(self):
        # compose a list of packets encoding the service table of this node
        services = []
        for service in self.serviceMap:
            capacity = 1 # TODO make local capacity variable
            services.append(makeNetworkServiceSharePacket(self.address, service, capacity))
        for service in self.serviceCapacityMap:
            capacityMap = self.serviceCapacityMap[service]
            for remoteAddress in capacityMap: # TODO sort by deccreasing capacity (first tx'd is highest capacity)
                # TODO filter expired or unroutable entries
                capacity = capacityMap[remoteAddress]
                services.append(makeNetworkServiceSharePacket(remoteAddress, service, capacity))
        return services

    def share_net_state(self):
        routes = self.export_routes()
        services = self.export_services()
        for channel in self.channels:
            for routePacket in routes:
                channel.send(routePacket)
            for servicePacket in services:
                channel.send(servicePacket)

    def close(self):
        self.stop = True
        for channel in self.channels:
            channel.close()

File: java/aln/src/main/java/org/biglittleidea/aln/Router.java
package org.biglittleidea.aln;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.HashMap;
import java.util.concurrent.ThreadLocalRandom;

public class Router {
    public Router(String address) {
        this.address = address;
    }

    public class ServiceNodeInfo {
        public String service; // name of the service
        public String address; // host of the service
        public String nextHop; // next routing node for address
        public Short capacity; // remote service capacity; zero means unavailable
        public String err; // parser error
    }

    // contextHandlerMap stores query response handers by contextID
    HashMap<Short, IPacketHandler> contextHandlerMap = new HashMap<Short, IPacketHandler>();

    // serviceHandlerMap maps local endpoint nodes to packet handlers
    HashMap<String, IPacketHandler> serviceHandlerMap = new HashMap<String, IPacketHandler>();

    public class RemoteNodeInfo {
        public String address; // target addres to communicate with
        public String nextHop; // next routing node for address
        public IChannel channel; // specific channel that is hosting next hop
        public short cost; // cost of using this route, generally a hop count
        public Date lastSeen; // routes should decay with a few missed updates

        public String err; // parser err msg
    }

    // map[address]RemoteNode
    HashMap<String, RemoteNodeInfo> remoteNodeMap = new HashMap<String, RemoteNodeInfo>();

    class NodeCapacity {
        short capacity;
        Date lastSeen;
    }

    // map[service][address]NodeCapacity
    HashMap<String, HashMap<String, NodeCapacity>> serviceCapacityMap = new HashMap<String, HashMap<String, NodeCapacity>>();

    // Router manages a set of channels and packet handlers
    protected String address = "";
    ArrayList<IChannel> channels = new ArrayList<>();

    ArrayList<String> serviceAddresses(String service) {
        ArrayList<String> addressList = new ArrayList<>();
        synchronized(this) {
            if (serviceHandlerMap.containsKey(service))
                addressList.add(this.address);

            if (serviceCapacityMap.containsKey(service)) {
                HashMap<String, NodeCapacity> addressToCapacityMap = serviceCapacityMap.get(service);
                for (String address : addressToCapacityMap.keySet()) {
                    addressList.add(address);
                }
            }
        }
        return addressList;
    }

    public String send(Packet p) {
        if (p.SourceAddress == null || p.SourceAddress.length() == 0) {
            p.SourceAddress = this.address;
        }
        if ((p.DestAddress == null || p.DestAddress.length() == 0) &&
                (p.Service != null && p.Service.length() > 0)) {
            ArrayList<String> addresses = this.serviceAddresses(p.Service);
            if (addresses.size() > 0) {
                for (int i = 1; i < addresses.size(); i++) {
                    Packet pc = p.copy();
                    pc.DestAddress = addresses.get(i);
                    this.send(pc);
                }
                p.DestAddress = addresses.get(0);
            } else {
                return String.format("no services of type %s yet discovered on the network", p.Service);
            }
        }

        if (p.DestAddress.equals(this.address)) {
            IPacketHandler handler;
            if (this.serviceHandlerMap.containsKey(p.Service)) {
                handler = this.serviceHandlerMap.get(p.Service);
                handler.onPacket(p);
            } else if (this.contextHandlerMap.containsKey(p.ContextID)) {
                handler = this.contextHandlerMap.get(p.ContextID);
                handler.onPacket(p);
            } else {
                return System.out.printf("service '%s' not registered\n", p.Service).toString();
            }
        } else if (p.NextAddress == null || p.NextAddress.equals(this.address) || p.NextAddress.length() == 0) {
            if (this.remoteNodeMap.containsKey(p.DestAddress)) {
                RemoteNodeInfo rni = this.remoteNodeMap.get(p.DestAddress);
                p.NextAddress = rni.nextHop;
                rni.channel.send(p);
            }
        } else {
            return "packet is unroutable; no action taken";
        }
        return null;
    }

    public short registerContextHandler(IPacketHandler handler) {
        short ctx;
        synchronized (this) {
            ctx = (short) ThreadLocalRandom.current().nextInt(1, 2 << 16);
            while (contextHandlerMap.containsKey(ctx))
                ctx = (short) ThreadLocalRandom.current().nextInt(1, 2 << 16);;
            contextHandlerMap.put(ctx, handler);
        }
        return ctx;
    }

    public void releaseContext(short ctx) {
        synchronized (this) {
            this.contextHandlerMap.remove(ctx);
        }
    }

    protected Packet composeNetRouteShare(String address, short cost) {
        Packet p = new Packet();
        p.Net = Packet.NetState.ROUTE;
        p.SourceAddress = this.address;
        p.Data = new byte[address.length() + 3];
        ByteBuffer buffer = ByteBuffer.allocate(Frame.BufferSize);
        buffer.put((byte) address.length());
        buffer.put(address.getBytes());
        buffer.put(Packet.writeUINT16(cost));
        buffer.rewind();
        buffer.get(p.Data);
        return p;
    }

    protected RemoteNodeInfo parseNetRouteShare(Packet packet) {
        RemoteNodeInfo info = new RemoteNodeInfo();
        if (packet.Net != Packet.NetState.ROUTE) {
            info.err = "parseNetworkRouteSharePacket: packet.NetState != NET_ROUTE";
            return info;
        }
        if (packet.Data == null || packet.Data.length == 0) {
            info.err = "parseNetworkRouteSharePacket: packet.Data is empty";
            return info;
        }

        byte[] data = packet.Data;
        byte addrSize = data[0];

        if (data.length != addrSize + 3) {
            info.err = String.format("parseNetworkRouteSharePacket: len: %d; exp: %d", data.length, addrSize + 3);
            return info;
        }

        int offset = 1;
        info.address = new String(Arrays.copyOfRange(data, offset, offset + addrSize));
        offset += addrSize;
        info.cost = Packet.readUINT16(data, offset);
        info.nextHop = packet.SourceAddress;
        return info;
    }

    protected Packet composeNetServiceShare(String address, String service, short capacity) {
        Packet p = new Packet();
        p.Net = Packet.NetState.SERVICE;
        p.SourceAddress = this.address;
        p.Data = new byte[this.address.length() + service.length() + 4];
        ByteBuffer buffer = ByteBuffer.allocate(Frame.BufferSize);
        buffer.put((byte) address.length());
        buffer.put(address.getBytes());
        buffer.put((byte) service.length());
        buffer.put(service.getBytes());
        buffer.put(Packet.writeUINT16(capacity));
        buffer.rewind();
        buffer.get(p.Data, 0, p.Data.length);
        return p;
    }

    protected ServiceNodeInfo parseNetServiceShare(Packet packet) {
        ServiceNodeInfo info = new ServiceNodeInfo();
        if (packet.Net != Packet.NetState.SERVICE) {
            info.err = "parseNetworkRouteSharePacket: packet.NetState != NET_SERVICE";
            return info;
        }
        if (packet.Data == null || packet.Data.length == 0) {
            info.err = "parseNetworkRouteSharePacket: packet.Data is empty";
            return info;
        }

        byte[] data = packet.Data;
        byte addrSize = data[0];
        int offset = 1;
        info.address = new String(Arrays.copyOfRange(data, offset, offset + addrSize));
        offset += addrSize;

        byte srvSize = data[offset];
        offset += 1;
        info.service = new String(Arrays.copyOfRange(data, offset, offset + srvSize));
        offset += srvSize;

        info.capacity = Packet.readUINT16(data, offset);
        return info;
    }

    protected Packet composeNetQuery() {
        Packet p = new Packet();
        p.Net = Packet.NetState.QUERY;
        return p;
    }

    protected void handleNetState(IChannel channel, Packet packet) {
        switch (packet.Net) {
            case Packet.NetState.ROUTE:
                // System.out.printf("router '%s' recv'd ROUTE update\n", address);
                // neighbor is sharing it's routing table
                RemoteNodeInfo info = parseNetRouteShare(packet);
                if (info.err != null) {
                    System.out.println(info.err);
                    return;
                }

                // msg := "NET_ROUTE [%s] remoteAddress:%s, nextHop:%s, cost:%d\n"
                // log.Printf(msg, r.address, remoteAddress, nextHop, cost)
                if (info.cost == 0) { // zero cost routes are removed
                    if (info.address.equals(this.address)) {
                        // fight the lie
                        try {
                            Thread.sleep(500);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        this.shareNetState();
                        return;
                    } else if (this.remoteNodeMap.containsKey(info.address)) {
                        // remove the route

                        RemoteNodeInfo localInfo = remoteNodeMap.get(info.address);
                        removeAddress(info.address);
                        for (IChannel ch : channels) {
                            if (ch != localInfo.channel) {
                                ch.send(packet);
                            }
                        }
                    }
                } else {
                    // add or update a route
                    RemoteNodeInfo localInfo;
                    synchronized (this) {
                        if (remoteNodeMap.containsKey(info.address)) {
                            localInfo = remoteNodeMap.get(info.address);
                        } else {
                            localInfo = new RemoteNodeInfo();
                            localInfo.address = info.address;
                            remoteNodeMap.put(info.address, localInfo);
                        }
                    }
                    localInfo.lastSeen = new Date();
                    if (info.cost < localInfo.cost || localInfo.cost == 0) {
                        localInfo.channel = channel;
                        localInfo.cost = info.cost;
                        localInfo.nextHop = info.nextHop;
                        Packet p = composeNetRouteShare(info.address, ++info.cost);
                        for (IChannel ch : channels) {
                            if (ch != channel) {
                                ch.send(p);
                            }
                        }
                    }
                }

                break;

            case Packet.NetState.SERVICE:
                // System.out.printf("router '%s' recv'd SERVICE update\n", address);
                ServiceNodeInfo serviceInfo = parseNetServiceShare(packet);
                if (serviceInfo.err != null) {
                    System.out.printf("error parsing net service: %s\n", serviceInfo.err);
                    return;
                }
                if (serviceInfo.address == this.address) {
                    break;
                }
                synchronized (this) {
                    HashMap<String, NodeCapacity> capacityMap;
                    if (!serviceCapacityMap.containsKey(serviceInfo.service)) {
                        capacityMap = new HashMap<>();
                        serviceCapacityMap.put(serviceInfo.service, capacityMap);
                    } else if (serviceInfo.capacity == 0) {
                        capacityMap = serviceCapacityMap.get(serviceInfo.service);
                        if (capacityMap.containsKey(serviceInfo.address))
                            capacityMap.remove(serviceInfo.address);
                        else // we have no record of this service at this address
                            return; // drop packet to avoid propagation loops
                    } else {
                        capacityMap = serviceCapacityMap.get(serviceInfo.service);
                        if (capacityMap.containsKey(serviceInfo.address) &&
                                capacityMap.get(serviceInfo.address).capacity == serviceInfo.capacity) {
                            return; // drop packet to avoid propagation loops
                        }
                    }
                    NodeCapacity nodeCapacity = new NodeCapacity();
                    nodeCapacity.capacity = serviceInfo.capacity;
                    nodeCapacity.lastSeen = new Date();
                    capacityMap.put(serviceInfo.address, nodeCapacity);
                }
                // forward the service capacity
                for (IChannel ch : channels) {
                    if (ch != channel) {
                        ch.send(packet);
                    }
                }
                break;

            case Packet.NetState.QUERY:
                // System.out.printf("router '%s' recv'd QUERY\n", address);
                for (Packet p : exportRouteTable())
                    channel.send(p);
                for (Packet p : exportServiceTable())
                    channel.send(p);
                break;
        }
    }

    public void addChannel(IChannel channel) {
        synchronized (this) {
            this.channels.add(channel);
        }

        channel.receive(new IPacketHandler() {
            @Override
            public void onPacket(Packet packet) {
                if (packet.Net != 0) {
                    handleNetState(channel, packet);
                } else {
                    send(packet);
                }
            }
        }, new IChannelCloseHandler() {
            @Override
            public void onChannelClosed(IChannel ch) {
                removeChannel(ch);
            }
        });
        // System.out.printf("router '%s' sending QUERY\n", address);

        channel.send(composeNetQuery()); // immediately query the new connection
    }

    public void removeChannel(IChannel ch) {
        synchronized (this) {
            channels.remove(ch);
            // bcast the loss of routes through the channel
            for (String address : remoteNodeMap.keySet()) {
                // System.out.printf("router:RemoveChannel address '%s'\n", address);
                RemoteNodeInfo nodeInfo = remoteNodeMap.get(address);
                if (nodeInfo.channel == ch) {
                    removeAddress(address);
                    for (IChannel c : channels) {
                        c.send(composeNetRouteShare(address, (short) 0));
                    }
                }
            }
        }
    }

    protected void removeAddress(String address) {
        synchronized (this) {
            remoteNodeMap.remove(address);
            for (HashMap<String, NodeCapacity> nlm : serviceCapacityMap.values()) {
                nlm.remove(address);
            }
        }
    }

    public void registerService(String service, IPacketHandler handler) {
        synchronized (this) {
            serviceHandlerMap.put(service, handler);
        }
    }

    public void unregisterService(String service) {
        synchronized (this) {
            serviceHandlerMap.remove(service);
        }
    }

    public Packet[] exportRouteTable() {
        Packet[] routes = new Packet[remoteNodeMap.size() + 1];
        int i = 0;
        routes[i++] = composeNetRouteShare(address, (short) 1);
        for (String address : remoteNodeMap.keySet()) {
            RemoteNodeInfo routeInfo = remoteNodeMap.get(address);
            routes[i++] = composeNetRouteShare(address, (short) (routeInfo.cost + 1));
        }
        return routes;
    }

    public Packet[] exportServiceTable() {
        int sz = serviceHandlerMap.size();
        for (HashMap<String, NodeCapacity> capacityMap : serviceCapacityMap.values()) {
            sz += capacityMap.size();
        }
        Packet[] services = new Packet[sz];
        int i = 0;
        for (String service : serviceHandlerMap.keySet()) {
            short capacity = 1; // TODO make local capacity variable
            services[i++] = composeNetServiceShare(this.address, service, capacity);
        }
        for (String service : serviceCapacityMap.keySet()) {
            HashMap<String, NodeCapacity> capacityMap = serviceCapacityMap.get(service);
            for (String address : capacityMap.keySet()) {
                NodeCapacity capacityInfo = capacityMap.get(address);
                services[i++] = composeNetServiceShare(address, service, capacityInfo.capacity);
            }
        }
        return services;
    }

    public void shareNetState() {
        Packet[] routes, services;
        synchronized (this) {
            routes = exportRouteTable();
            services = exportServiceTable();
        }
        for (IChannel ch : channels) {
            for (Packet p : routes)
                ch.send(p);
            for (Packet p : services)
                ch.send(p);
        }
    }

    public int numChannels() {
        synchronized (this) {
            return channels.size();
        }
    }

}
